File: src/brs/web/api/http/handler/GenerateDeeplink.java
package brs.web.api.http.handler;

import brs.deeplink.DeeplinkGenerator;
import brs.util.Convert;
import brs.util.StringUtils;
import brs.web.api.http.ApiServlet;
import brs.web.api.http.common.LegacyDocTag;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.servlet.http.HttpServletRequest;
import java.io.UnsupportedEncodingException;

import static brs.web.api.http.common.JSONResponses.*;
import static brs.web.api.http.common.Parameters.*;

public class GenerateDeeplink extends ApiServlet.JsonRequestHandler {

  private final Logger logger = LoggerFactory.getLogger(GenerateDeeplink.class);
  public static final GenerateDeeplink instance = new GenerateDeeplink();


  private GenerateDeeplink() {
    super(new LegacyDocTag[]{LegacyDocTag.UTILS}, ACTION_PARAMETER, PAYLOAD_PARAMETER);
  }

  @Override
  public JsonElement processRequest(HttpServletRequest req) {
    try {

      final String action = Convert.emptyToNull(req.getParameter(ACTION_PARAMETER));
      final String payload = Convert.emptyToNull(req.getParameter(PAYLOAD_PARAMETER));

      if (StringUtils.isEmpty(action) && !StringUtils.isEmpty(payload)) {
        return PAYLOAD_WITHOUT_ACTION;
      }

      DeeplinkGenerator deeplinkGenerator = new DeeplinkGenerator();
      String deepLink = deeplinkGenerator.generateDeepLink(action, payload);
      JsonObject response = new JsonObject();
      response.addProperty("link", deepLink);
      return response;

    } catch (IllegalArgumentException e) {
      logger.error("Problem with arguments", e);
      return incorrect("arguments", e.getMessage());
    } catch (UnsupportedEncodingException e) {
      throw new RuntimeException(e);
    }
  }
}

File: src/main/kotlin/brs/api/http/GenerateDeeplink.kt
package brs.api.http

import brs.api.http.common.JSONResponses.MISSING_DOMAIN
import brs.api.http.common.JSONResponses.PAYLOAD_WITHOUT_ACTION
import brs.api.http.common.JSONResponses.incorrect
import brs.api.http.common.Parameters.ACTION_PARAMETER
import brs.api.http.common.Parameters.DOMAIN_PARAMETER
import brs.api.http.common.Parameters.PAYLOAD_PARAMETER
import brs.api.http.common.ResultFields.DEEPLINK_RESPONSE
import brs.services.DeeplinkGeneratorService
import brs.util.convert.emptyToNull
import brs.util.jetty.get
import brs.util.logging.safeDebug
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import org.slf4j.LoggerFactory
import javax.servlet.http.HttpServletRequest

internal class GenerateDeeplink(private val deeplinkGeneratorService: DeeplinkGeneratorService) :
    APIServlet.JsonRequestHandler(arrayOf(APITag.UTILS), DOMAIN_PARAMETER, ACTION_PARAMETER, PAYLOAD_PARAMETER) {
    override fun processRequest(request: HttpServletRequest): JsonElement {
        try {
            val domain = request[DOMAIN_PARAMETER].emptyToNull()
            if (domain.isNullOrEmpty()) {
                return MISSING_DOMAIN
            }

            val action = request[ACTION_PARAMETER].emptyToNull()
            val payload = request[PAYLOAD_PARAMETER].emptyToNull()

            if (action.isNullOrEmpty() && !payload.isNullOrEmpty()) {
                return PAYLOAD_WITHOUT_ACTION
            }

            val deepLink = deeplinkGeneratorService.generateDeepLink(domain, action, payload)
            val response = JsonObject()
            response.addProperty(DEEPLINK_RESPONSE, deepLink)
            return response
        } catch (e: IllegalArgumentException) {
            logger.safeDebug(e) { "Problem with arguments" }
            return incorrect("arguments", e.message ?: "")
        }
    }

    companion object {
        private val logger = LoggerFactory.getLogger(GenerateDeeplink::class.java)
    }
}
